_extras/refactor-3-code-abstractions.md

```diff
@@ -0,0 +1,482 @@
+
+```python
+class Rectangle:
+  def __init__(self, width, height):
+    self.width = width
+    self.height = height
+  def get_area(self):
+    return self.width * self.height
+```
+
+Like `Circle`, this class provides the `get_area()` method.
+The method takes the same number of parameters (none), and returns a number.
+However, the implementation is different. This is one type of *polymorphism*.
+
+The word "polymorphism" means "many forms", and in programming it refers to 
+methods/functions/operators with the same name that can be executed on many objects or classes.
+
+Using our `Circle` and `Rectangle` classes, we can create a list of different shapes and iterate 
+through the list to find their total surface area as follows:
+
+```python
+my_circle = Circle(radius=10)
+my_rectangle = Rectangle(width=5, height=3)
+my_shapes = [my_circle, my_rectangle]
+total_area = sum(shape.get_area() for shape in my_shapes)
+```
+
+Note that we have not created a common superclass or linked the classes `Circle` and `Rectangle` 
+together in any way. It is possible due to polymorphism. 
+You could also say that, when we are calculating the total surface area, 
+the method for calculating the area of each shape is abstracted away to the relevant class.
+
+How can polymorphism be useful in our software project? 
+For example, we can replace our `CSVDataSource` with another class that reads a totally 
+different file format (e.g. JSON instead of CSV), or reads from an external service or database
+All of these changes can be now be made without changing the analysis function as we have decoupled 
+the process of data loading from the data analysis earlier.
+Conversely, if we wanted to write a new analysis function, we could support any of these 
+data sources with no extra work.
+
+> ## Exercise: Add an Additional DataSource
+> Create another class that supports loading catchment data from JSON files, with the 
+> appropriate `load_catchment_data()` method.
+> There is a function in `models.py` that loads from JSON in the following format:
+> ```json
+>[
+>    {
+>        "Site": "FP35",
+>        "Site Name": "Lower Wraxall Farm",
+>        "Date": "01/12/2008 23:00",
+>        "Rainfall (mm)": 0.0
+>    },
+>    {
+>        "Site": "FP35",
+>        "Site Name": "Lower Wraxall Farm",
+>        "Date": "01/12/2008 23:15",
+>        "Rainfall (mm)": 0.0
+>    }
+> ]
+> ```
+> Finally, at run time construct an appropriate instance based on the file extension.
+>> ## Solution
+>> The new class could look something like:
+>> ```python
+>> class JSONDataSource:
+>>     """
+>>     Loads patient data with catchment values from JSON files within a specified folder.
+>>     """
+>>     def __init__(self, dir_path):
+>>         self.dir_path = dir_path
+>>
+>>     def load_catchment_data(self):
+>>         data_file_paths = glob.glob(os.path.join(self.dir_path, 'rain_data_2015*.json'))
```

Comment:
So I think these lessons are looking really good, my one niggle is with the arguments for how the files are being read - the controller still asks for 'Input CSV(s) containing measurement data', and is still only looking at the directory path and whether it ends with .json or .csv. I'm not sure how best to fix it, I think there should at least be some suggestion of how to fix this, or maybe some additional optional arguments in the next section (say -j puts it in json mode)?

Reply:
I agree that it's poor practice to look at the file endings for this work. We perhaps could include an exercise in the last lesson, for the students to replace this file ending check with a flag instead? I'll look at the options for this.

Reply:
But, for now, shall we merge this material? (and I'll add that exercise as another PR)

Reply:
Yeah, this is more of a comment than a requirement, happy for it to be merged as is